src/transformers/modeling_utils.py

```diff
@@ -1485,13 +1484,12 @@ def load(module: nn.Module, prefix=""):
 
         # Make sure we are able to load base models as well as derived models (with heads)
         start_prefix = ""
-        model_to_load = model
         if not hasattr(model, cls.base_model_prefix) and has_prefix_module:
             start_prefix = cls.base_model_prefix + "."
+        load(model, prefix=start_prefix)
```

Comment:
Why is this added here? In the case the following test is positive, we end up loading the model twice.

Reply:
It seems that `from_pretrained` ignores the HEAD weights in some models, e.g.:

```
python -c 'from transformers import AutoModelForMaskedLM; model = AutoModelForMaskedLM.from_pretrained("Narsil/tiny-distilbert"); print(model)'

Some weights of the model checkpoint at Narsil/tiny-distilbert were not used when initializing DistilBertForMaskedLM: ['qa_outputs.weight', 'classifier.weight', 'pre_classifier.bias', 'qa_outputs.bias', 'classifier.bias', 'pre_classifier.weight']
[...]
```

Reply:
ah, @Narsil, this is because you used the wrong module, this works correctly:
```
python -c 'from transformers import DistilBertForSequenceClassification; model = DistilBertForSequenceClassification.from_pretrained("Narsil/tiny-distilbert"); print(model)'
```

Reply:
This is exactly what failed actually. Bit of context, I used this script: https://gist.github.com/Narsil/7198efc17ad5c9bc4f711ffb0e0e3758 based on the original from @LysandreJik .

I realized there was an issue where the weights where incorrect (indeed we can't have 1 model for every possible head since at least ForMultipleChoice and ForSequenceClassification share the module `classifier` yet have different shapes (dim, 1) vs (dim, num_labels)).

I don't really know, but I ended up with a wrong model without seeing any error. Can't reproduce today for some reason.

Reply:
(When size mismatched, instead of loading the head or crashing, I ended with a random head in the pipeline leading to random results)

Reply:
Okay, to reproduce:

```python
DistilBertForSequenceClassification.from_pretrained("Narsil/tiny-distilbert-sequence-classification").classifier.weight
```
`python -c 'from transformers import *; print(DistilBertForSequenceClassification.from_pretrained("Narsil/tiny-distilbert-sequence-classification").classifier.weight)'`  (for the lazy like me)

Run that line a couple of times, and you will have different weights, even though the state_dict actually contains the head.

Reply:
(And there is no log whatsoever)

Reply:
It looks like that particular model has not been saved properly: its state_dict contains keys that do not begin `distilbert`, as if it was an model without head, but still contains keys for a head. That's what makes the `from_pretrained` method confused but the problem is not in that method, it's this model state dict that is corrupted.

I double-checked but saving a `DistilBertForSequenceClassification` with either `save_pretrained` or `torch.save(model.state_dict())` will produce a usable state dict so I'm not sure how you obtained this one.

Reply:
Ok, makes sense, I used the script made by @LysandreJik but I guess there was an issue, I'll fix it there then.

Reply:
@sgugger the fact that it's not saying that it's missing keys is a bug though, no ? (The script had checks in place to make sure there were no missing keys, but nothing is reported)

Reply:
```python
from transformers import *

model = DistilBertForSequenceClassification.from_pretrained(None, config=DistilBertConfig(), state_dict = {})
model.state_dict().keys()    # "classifier.weight" and not "distilbert.classifier.weight"
model.save_pretrained("tmp")  # Same in the pytorch dict.
```
Let's move this discussion online so it's an easier back and forth :)

Reply:
Yes, the wrong state dict definitely confused the `from_pretrained` function but there is a bug to fix indeed!